extensions/amp-analytics/0.1/variables.js

```diff
@@ -0,0 +1,276 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev, user} from '../../../src/log';
+import {fromClass} from '../../../src/service';
+import {isArray} from '../../../src/types';
+import {cryptoFor} from '../../../src/crypto';
+
+const TAG = 'Analytics.Variables';
+
+/** @typedef {function(...?):(?|!Promise<?>)} */
+let VariableFilterDef;
+
+/**
+ * @param {!string} str
+ * @param {Number} s
+ * @param {Number} l
+ * @return {string}
+ */
+function substrFilter(str, s, l) {
+  const start = Number(s);
+  let length = str.length;
+  user().assertNumber(start,
```

Comment:
changed. Was thinking of putting range checks as well but since substr accepts arbitrary values, didn't add them here.